tests/unit/fem/ptest_tdofs.cpp

```diff
@@ -0,0 +1,1012 @@
+
+      delete mesh;
+      mesh = nullptr;
+
+      refineMesh(pmesh, NUM_PAR_REF_LEVELS);
+   }
+
+   SECTION("Testing true dofs, without curvature")
+   {
+      TDOFTester  tester(pmesh, VERBOSE_OUTPUT);
+      tester.testVertexTDofs();
+      tester.testEdgeTDofs();
+      tester.testFaceTDofs();
+      tester.testElementTDofs();
+
+      tester.testBoundaryElementTDofs();
+   }
+
+   SECTION("Testing true dofs, w/ curvature order 1")
+   {
+      pmesh->SetCurvature(1);
+
+      TDOFTester  tester(pmesh, VERBOSE_OUTPUT);
+      tester.testVertexTDofs();
+      tester.testEdgeTDofs();
+      tester.testFaceTDofs();
+      tester.testElementTDofs();
+
+      tester.testBoundaryElementTDofs();
+   }
+
+   SECTION("Testing true dofs, w/ curvature order 2")
+   {
+      pmesh->SetCurvature(2);
+
+      TDOFTester  tester(pmesh, VERBOSE_OUTPUT);
+      tester.testVertexTDofs();
+      tester.testEdgeTDofs();
+      tester.testFaceTDofs();
+      tester.testElementTDofs();
+
+      tester.testBoundaryElementTDofs();
+   }
+
+   delete pmesh;
+}
+
+TEST_CASE("NC Quad mesh tdofs",
+          "[ParFESpace]"
+          "[NCMesh]"
+          "[Hypre]"
+          "[TDofs]")
+{
+   int rank;
+   MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+   MPRINT_MASTER(rank,
+                 "=== Checking TDOFs for Non-conforming Quad mesh ===");
+
+   const double refine_prob = .5;
+
+   const int attrib = 1;
+   mfem::Mesh* mesh = nullptr;
+   mfem::ParMesh* pmesh = nullptr;
+
+   // Build a simple single element quad mesh
+   const int dim = 2, nv = 4, ne = 1, nb = 0, sdim = 2;
+   {
+      mesh = new mfem::Mesh(dim, nv, ne, nb, sdim);
+
+      mesh->AddVertex(mfem::Vertex(0.,0.)());
+      mesh->AddVertex(mfem::Vertex(1.,0.)());
+      mesh->AddVertex(mfem::Vertex(1.,1.)());
+      mesh->AddVertex(mfem::Vertex(0.,1.)());
+
+      int idx[4] = {0,1,2,3};
+      mesh->AddElement(new mfem::Quadrilateral(idx, attrib));
+
+      mesh->FinalizeMesh();
+
+      CHECK( mesh->GetNV() == nv);
+      CHECK( mesh->GetNE() == ne);
+      CHECK( mesh->GetNBE() == 4);
+      CHECK( mesh->Dimension() == dim);
+      CHECK( mesh->SpaceDimension() == sdim);
+
+      mesh->SetCurvature(2);
+      mesh->EnsureNCMesh();
+   }
+
+   // refine serial mesh
+   {
+      // Refine the mesh once before random refinement
+      refineMesh(mesh, 1);
+
+      for (int i=0; i < NUM_SER_REF_LEVELS; ++i)
+      {
+         mesh->RandomRefinement(refine_prob);
+      }
+   }
+
+   MPRINT_MASTER(rank, "Details for serial mesh, after random NC refinement:\n"
+                 << getMeshStats(mesh));
+
+   // Create and refine parallel mesh
+   {
+      pmesh = new mfem::ParMesh(MPI_COMM_WORLD, *mesh);
+
+      delete mesh;
+      mesh = nullptr;
+
+      for (int i=0; i < NUM_PAR_REF_LEVELS; ++i)
+      {
+         pmesh->RandomRefinement(refine_prob);
+      }
+   }
+
+   SECTION("Testing true dofs for nc mesh")
+   {
+      TDOFTester  tester(pmesh, VERBOSE_OUTPUT);
+
+      /// WARNING:
```

Comment:
For (2D) NC meshes, I was able to generate the element tdofs and the boundary element tdofs, but not the vertex or edge tdofs due to the issue described above.